### PR TITLE
Dark collection state

### PR DESCRIPTION
**Pull request by JAAlvarado-Montes**
Hi all,

This is the first PR for the dark-collection mode of Huntsman, saved as `states/huntsman/taking_darks.py`. I based this mode on the housekeeping.py mode of Huntsman, and the modified scripts are `scheduler/observation.py` and `huntsman/observatory.py`. For the first one a new class `DarkObservation` has been included, and for the second one a new method was created `take_dark_fields` and a new private method called `_create_dark_observation`. All the new definitions have been based on previously existent code within the same scripts. I know there are many things to improve, so this is the first PR to start pulling off this new mode of Huntsman.

Any comment or suggestion is welcome.

Cheers,
Jaime.

### CODE REVIEW

File: states/huntsman/taking_darks.py
Comment (lspitler) [inline states/huntsman/taking_darks.py]: Should the mount be parked before darks are taken?

File: states/huntsman/taking_darks.py
Comment (lspitler) [inline states/huntsman/taking_darks.py]:
```suggestion
    """
```

File: states/huntsman/taking_darks.py
Comment (lspitler) [inline states/huntsman/taking_darks.py:20]: I think you need to call `take_dark_fields` somewhere here, right?

File: huntsman/observatory.py
Comment (lspitler) [inline huntsman/observatory.py]: Why do you have this parameter?

File: huntsman/observatory.py
Comment (lspitler) [inline huntsman/observatory.py]: this parameter isn't used:
```suggestion
                         
```

File: huntsman/observatory.py
Comment (lspitler) [inline huntsman/observatory.py]: Looks like you're hard-coding 1 second exposures here for some reason. Shouldn't this be given as a arg to `take_dark_fields`?

File: huntsman/observatory.py
Comment (lspitler) [inline huntsman/observatory.py]: How do you get out of this loop? I don't see a `break`

File: huntsman/observatory.py
Comment (lspitler) [inline huntsman/observatory.py]: remove these lines?

File: huntsman/observatory.py
Comment (lspitler) [inline huntsman/observatory.py]: Don't think this is needed. So `max_exptime` is also not needed?

File: huntsman/observatory.py
Comment (lspitler) [inline huntsman/observatory.py]:
```suggestion
                time.sleep(wait_interval)
```

As in `pointing.py`

File: huntsman/observatory.py
Comment (lspitler) [inline huntsman/observatory.py]: Again, you're hardwiring in an exposure time.

**Comment by AnthonyHorton** [comment]
@JAAlvarado-Montes When you get a chance please resolve the merge conflicts. The tests won't run until those are fixed.

**Review by lspitler — CHANGES_REQUESTED** [review]
@JAAlvarado-Montes good first step. Some changes and questions for you here. Let me know when you've addressed them or you want to chat about them. Thanks

**Comment by AnthonyHorton** [comment]
@JAAlvarado-Montes Have you tested this in simulation? Based on a quick look I don't think it would actually work. Your new state doesn't call the new HuntsmanObservatory method for taking darks at all, and if it did the HuntsmanObservatory would keep taking darks forever.

**Comment by lspitler** [comment]
@JAAlvarado-Montes are you able to progress this so I can do another review before I leave on the 9th?

File: huntsman/observatory.py
Comment (wtgee) [reply huntsman/observatory.py]: I also don't see a `break` :)

**Review by wtgee — CHANGES_REQUESTED** [review]
Looks like this has mostly been hashed out and a copy of the diterhed except for the identified lack of `break` statement.

**Comment by JAAlvarado-Montes** [comment]
Hi @lspitler, @wtgee, and @AnthonyHorton,

I have addressed all of the changes and suggestions you made for this PR. The three modified scripts are taking_darks.py (state), observatory.py containing the function `take_dark_fieds` and private method `_create_dark_observation`; and observation.py which has the subclass `DarkObservation` used by the aforementioned function. 

Something I'm not sure how to do (and that I haven't done yet) is how to decide when the taking_darks state should be triggered. I think the two important conditions to be met would be `if pocs.is_dark and not pocs.can_observe:...` but I'm not certain whether I'm missing something or if those two conditions are right. Also, where would these conditions go? I assume that they cannot go in the taking_darks state, so I think they should go to a different state (maybe observing.py?) that triggers the "dark" state, but I want to be sure before doing this step.

If you have some chance, please review this PR and let me know if you have any feedback. It will be really appreciated.

Cheers,
Jaime.

**Comment by wtgee** [comment]
> Something I'm not sure how to do (and that I haven't done yet) is how to decide when the taking_darks state should be triggered. I think the two important conditions to be met would be `if pocs.is_dark and not pocs.can_observe:...` but I'm not certain whether I'm missing something or if those two conditions are right. Also, where would these conditions go? I assume that they cannot go in the taking_darks state, so I think they should go to a different state (maybe observing.py?) that triggers the "dark" state, but I want to be sure before doing this step.
> 
> If you have some chance, please review this PR and let me know if you have any feedback. It will be really appreciated.

I can explain how the state setup works and give a PR review later today.

**Comment by JAAlvarado-Montes** [comment]
Great @wtgee, I'll be attentive to it. 

Cheers!

File: states/huntsman/taking_darks.py
Comment (wtgee) [inline states/huntsman/taking_darks.py]: The mount has to be parked to do these darks, so the _previous_ state will be `parked`, but the `next_state` should probably be `housekeeping` (or `sleeping`).

Comment (wtgee) [reply states/huntsman/taking_darks.py]: Following up on this, that also means that the logic to get **to** the `taking_darks` state should go in `parked`. 

It should be changed to something like:

```py
# states/huntsman/parked.py

# Psuedo-code below
if is_dark and weather_is_bad:
    pocs.next_state = 'taking_darks'
else:
    pocs.next_state = 'housekeeping'
```

It might be more complex than that, depending.

Alternatively, we could consider the act of taking darks to be part of `housekeeping` itself, but we probably best to keep it a little separate.

File: states/huntsman/parked.py
Comment (wtgee) [inline states/huntsman/parked.py]:
```suggestion
    if pocs.is_dark(horizon='flat') and pocs.is_weather_safe is False:
```

We don't have a `weather_is_bad`, I was just being lazy in my earlier comment (that's why it said it was psuedo-code).

Comment (JAAlvarado-Montes) [reply states/huntsman/parked.py]: OK, thanks for the input, I'll change this

**Comment by JAAlvarado-Montes** [comment]
Hi @AnthonyHorton, @wtgee, and @danjampro,

I've modified the taking_darks.py script to have a `while` loop that meets the two necessary conditions for taking darks. However, I'd like you to give me some input on this part as I'm not sure whether the syntax I'm following is correct. I didn't put an explicit `break` statement because as far as I'm concerned that should be implicit(?). I mean, once those two conditions are not met anymore the `while` loop should be broken automatically and an `Exception` raised (which is what I did). After that `Exception` the `housekeeping` state is triggered.

Once I address this, I want to continue and do the required test for this state. Question: Should that test be placed in `huntsman/tests`? Or should it live in a different location? I ask since I don't see any other tests regarding huntsman states.

Thanks in advance for taking some time to review this, I appreciate it.

Cheers,
Jaime.

File: huntsman/tests/test_huntsman.py
Comment (danjampro) [inline huntsman/tests/test_huntsman.py]:
```suggestion
    pocs.is_dark(horizon="flat") = True
```

File: huntsman/tests/test_huntsman.py
Comment (danjampro) [inline huntsman/tests/test_huntsman.py]:
```suggestion
    pocs.is_weather_safe = False
```

File: huntsman/tests/test_huntsman.py
Comment (danjampro) [reply huntsman/tests/test_huntsman.py]: Are you trying to setup the conditions for the dark state here? If so, I think = should be used.

File: states/huntsman/taking_darks.py
Comment (danjampro) [inline states/huntsman/taking_darks.py]: I'm not sure the error message is appropriate. An exception won't be raised if the while condition is not met. Maybe:
```suggestion
        pocs.logger.warning("Problem encountered while taking darks: {}".format(e))
```

File: states/huntsman/taking_darks.py
Comment (danjampro) [inline states/huntsman/taking_darks.py:20]: I think there should be a break statement in here, otherwise (e.g. on a rainy night) we would end up with 12 hours worth of darks. Either that, or you could change the `while` statement to an `if` statement e.g.:

```
if pocs.is_dark(horizon="flat") and pocs.is_weather_safe is False:
    <code here>
else:
    pocs.logger.info("Conditions for dark fields are not met")
```

File: huntsman/observatory.py
Comment (danjampro) [inline huntsman/observatory.py]: Should this not be a loop over exposure times? Why is `exptimes[cam_name][-1]` used?

File: huntsman/observatory.py
Comment (danjampro) [inline huntsman/observatory.py]: What happens if `exptimes_list` is empty? Do / should we have a predefined sequence of dark exposure times?

**Review by danjampro — CHANGES_REQUESTED** [review]
Hi @JAAlvarado-Montes, I've added a few questions, suggestions and discussion points to the PR. I have two main concerns at the moment:

1) It doesn't seem that you are looping over the exposure times in `exptimes_list`, but rather doing one exposure using the last entry.

2) There is no way the dark collection state will stop if it is dark and there is bad weather. I think this will lead to far too many darks being taken on some nights (and will probably cause your test to hang indefinitely). I'm not sure the `try, while, except` structure you have used in `on_enter` will do what you intended it to.

File: huntsman/tests/test_huntsman.py
Comment (JAAlvarado-Montes) [reply huntsman/tests/test_huntsman.py]: Yes, I am setting up the conditions for the dark state. I tried using `=` but got an error which says "can't assign to function call". However, I've changed the way to set up (simulate) those conditions. Please let me know what you think about that. Thanks!

File: huntsman/tests/test_huntsman.py
Comment (JAAlvarado-Montes) [reply huntsman/tests/test_huntsman.py]: Same here (see the previous answer).

File: states/huntsman/taking_darks.py
Comment (JAAlvarado-Montes) [reply states/huntsman/taking_darks.py]: Changed this to the suggested message. However, it's been moved to the `parked` state where the conditions for triggering `taking_darks` are.

File: states/huntsman/taking_darks.py
Comment (JAAlvarado-Montes) [reply states/huntsman/taking_darks.py:20]: I noticed that I put the conditions to trigger the `taking_darks` state inside the `parked` state using an `if` statement as you suggested. For that reason, I have removed them from the `taking_darks` state since if this state is triggered it's because the conditions were already met in `parked`, so it's not necessary to repeat them inside `taking_darks`. `taking_darks` now has only a `try` and `except` statement.

File: huntsman/observatory.py
Comment (JAAlvarado-Montes) [reply huntsman/observatory.py]: The [-1] was a typing error. Thanks for this!

File: huntsman/observatory.py
Comment (JAAlvarado-Montes) [reply huntsman/observatory.py]: Well, I reckon there should always be some targets in the config file, and the exposure times are taken from there to fill the list that is passed into the function `take_dark_fields`. So the only way for that list to be empty is that the targets config file is totally empty. If that happens, then `exptimes_list` would be empty. But I'm not sure if that's feasible since we always start the night with some targets in that file (and hence their exposure times), don't we? 
If there's a possibility that such a file be empty for any reason, then as you pointed out we should have a predefined list of exposure times. It totally depends on the config file for the targets.

**Comment by JAAlvarado-Montes** [comment]
Hello @AnthonyHorton, @danjampro, and @wtgee,

This PR is ready for a "final" review. On the one hand, I have written two tests to `test_huntsman.py`, namely  `test_entering_darks_state` and `test_darks_collection_simulator`. The former is to test the conditions for entering the `taking_darks` state, and the latter is for testing the method that takes darks. For testing I'm using the docker test image environment and both tests are passing without any conflict. `test_entering_darks_state` simply simulates the conditions needed for taking darks, which are when the sky is dark and the weather is not safe. `test_darks_collection_simulator` does simply the same than the state would do when taking darks, but using simulated cameras. It searches into the `pocs.observatory.scheduler.fields_list` to collect all of the exposure times that have been used before when observing, and create a list of exposure times (without repeating) which will be parsed to the `take_dark_fields` method in `observatory.py`

On the other hand, the `take_dark_fields` method in `observatory.py` is designed so that given a list of exposure times it will take a sequence of 10 darks per exposure time (i.e. by default, but can be modified) exposing all of the cameras at the same time and waiting for all of them to finish before taking the next exposure in the sequence. Once a sequence is done, it continues with the next exposure time on the list and so on.

I have tested `test_darks_collection_simulator` for different numbers of exposure times in `pocs.observatory.scheduler.fields_list` and it is working perfectly.  However, your recommendations are very welcome and I would be glad to add any change you consider important if you spot some things I have not, so I appreciate your input.

Cheers,
Jaime.

File: huntsman/scheduler/observation.py
Comment (wtgee) [inline huntsman/scheduler/observation.py]: You should be able to just pass the `position` directly

```suggestion
        dark_field = Field('Dark Field', position)
```

File: huntsman/scheduler/observation.py
Comment (wtgee) [inline huntsman/scheduler/observation.py]: This could use a docstring with some explanation about how this `exptime` is different.

File: huntsman/scheduler/observation.py
Comment (wtgee) [inline huntsman/scheduler/observation.py]: This whole class could probably go in its own file.

File: huntsman/scheduler/observation.py
Comment (wtgee) [inline huntsman/scheduler/observation.py]: Same about the docstring.

File: huntsman/scheduler/observation.py
Comment (wtgee) [inline huntsman/scheduler/observation.py]: We might want to build in an actual exposure index that increments when taking pictures but I guess I don't see any need.

But `_exp_index` doesn't need the `_` ;)

File: huntsman/scheduler/observation.py
Comment (wtgee) [inline huntsman/scheduler/observation.py]: Just a reminder you can get in the habit of:

```suggestion
        return f"DarkObservation: {self._field}: {self._exptime}"
```

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]: This needs to be cleaned up to match.

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]: `cameras_all` is a bad name because it is not necessarily always all cameras (such as in this line here).

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]: Good examples of the f-strings being better.
```suggestion
        self.logger.debug(f"Going to take {n_darks} dark fields for each of these exposure times {exptimes}")
```

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]: I want to say let's remove this and try to remember in case we see weird things. But ideally we don't have a line we don't understand.

```suggestion
```

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]: Slightly more pythonic to just assume and ignore exception.
```suggestion
            with suppress(AttributeError):
                exptime = exptime.to_value(u.second)
```

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]:
```suggestion
                self.logger.debug(f'Dark sequence #{num} of exposure time {exptime}s')
```

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]: This statement isn't quite true yet as we haven't yet started exposing.  Could also change to f-string.

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]: Better to use `os.path.join` here:

```suggestion
                    filename = os.path.join(
                        image_dir,
                        'darks',
                        camera.uid,
                        dark_obs.seq_time,
                        f'dark_{num:02d}_{exptime:.02f}s.{camera.file_extension}')
```

Although I'm not sure we want to continue to hard-code the `darks` directory.  Also the filename is a little weird with the exptime so you might want to alter to waht you want.

File: huntsman/observatory.py
Comment (wtgee) [inline huntsman/observatory.py]: I don't think you want to block here as it waits below.

```suggestion
                        blocking=False
```

**Review by wtgee — COMMENTED** [review]
Looking good. I did a first quick pass because there are some changes in it and tests aren't passing right now. 

So sorry, probably not the "final" version quite yet. :wink:

File: huntsman/observatory.py
Comment (danjampro) [inline huntsman/observatory.py]: What are the units of `wait_interval`? Probably seconds - but good to be specific.

File: states/huntsman/parked.py
Comment (danjampro) [inline states/huntsman/parked.py:10]: O.K. so we are using the state `taking_darks` to actually take the darks. The problem is I can't see this state listed in the state table (`resources/state_table/huntsman.yaml`), which needs to declare the specify the allowable transitions in/out of `taking_darks`.

File: huntsman/observatory.py
Comment (danjampro) [inline huntsman/observatory.py]: This description is not quite right here because this function does not look up the exposure times in the target list. Instead, this function expects an `exptimes` argument. Maybe a better description would be something like:

"Take n_darks dark frames for each exposure time specified, for each camera."

File: huntsman/tests/test_huntsman.py
Comment (danjampro) [inline huntsman/tests/test_huntsman.py]: This is not testing whether it can enter the `taking_darks` state (I don't actually think it can at the moment because there is no mention of `taking_darks` in the state table, see comment below).

So first you assert that the conditions are right to enter the dark state. So you need it to be a) dark, b) weather not safe and c) in the correct state to transition into `taking_darks`. It looks like this should be `parked` from the rest of your code, not `sleeping`. You will then need something like:

```
        assert pocs.next_state == "taking_darks"
        pocs.goto_next_state()
        assert pocs.state == "taking_darks"
``` 

for this to actually test whether it successfully enters the state. Finally, I think this test would be better suited to being in `tests/test_states.py` rather than `tests/test_huntsman.py`. I would suggest taking a look at some of the other tests in `tests/test_states.py` if this is not clear.

File: huntsman/tests/test_huntsman.py
Comment (danjampro) [inline huntsman/tests/test_huntsman.py:236]: - It would be nice to include a short description of what this test is trying to achieve.
- What is the purpose of `return False` at the end of this function?

File: huntsman/tests/test_huntsman.py
Comment (danjampro) [inline huntsman/tests/test_huntsman.py:260]: I think you need to actually check whether the FITS files are actually created at the end of this test. You can use something like 
`assert(os.path.isfile(filename))`, or you could count the number of FITS files in the relevant directories etc.

**Review by danjampro — CHANGES_REQUESTED** [review]
Hi @JAAlvarado-Montes, thanks for your work on this so far. I have two concerns that should be addressed before we merge this:

1) There is no way for the state machine to transition in/out of the `taking_darks` state because it is not in the state resource table (see comments below).

2) The test to check if the state machine successfully enters the `taking_darks` state does not work. If it did, it would pick up on the above. Simply changing the state like `pocs.next_state="taking_darks"` directly is not what we want to test, but rather we should check if the state machine naturally transitions to `taking_darks` given the right conditions. See below for some suggestions on this.

The other thing is that our tests now fail because of the Travis build (not related to your PR). @wtgee , thoughts on this?

File: huntsman/tests/test_huntsman.py
Comment (JAAlvarado-Montes) [reply huntsman/tests/test_huntsman.py:260]: I'm mainly asserting two things now: 1) To check whether all of the created files are fits, and 2) verify that the number of created fits files equals the theoretical (expected) number of darks.

**Comment by JAAlvarado-Montes** [comment]
Hello @danjampro, @wtgee, and @AnthonyHorton  

I have addressed all of @wtgee's and @danjampro's comments. As it is now, tests are passing and **we are indeed testing** that given the required conditions, namely bad weather and dark sky, the state machine goes from `parked` to `taking_darks`. I have added the `taking_darks` state to the states table along with its allowable transitions and migrated `test_entering_darks_state()` from `tests/test_huntsman.py` to `tests/test_states.py`

Please let me know what you think of this PR's current state, and as always, any thought or recommendation for moving it forward is very welcome.

Cheers,
Jaime A.

**Comment by JAAlvarado-Montes** [comment]
Hello @danjampro, @wtgee, and @AnthonyHorton

I have addressed all of @wtgee's and @danjampro's comments. As it is now, tests are passing and we are indeed testing that given the required conditions, namely bad weather and dark sky, the state machine goes from parked to taking_darks. I have added the taking_darks state to the states table along with its allowable transitions and migrated test_entering_darks_state() from tests/test_huntsman.py to tests/test_states.py

Please let me know what you think of this PR's current state, and as always, any thought or recommendation for moving it forward is very welcome.

Cheers,
Jaime A.

File: resources/state_table/huntsman.yaml
Comment (danjampro) [inline resources/state_table/huntsman.yaml]: I think the trigger to enter the `taking_darks` state should have its own trigger. Something like `trigger: take_darks`.

**Review by danjampro — COMMENTED** [review]
@JAAlvarado-Montes  Looks good to me. I left one comment about changing the `trigger` for entering the new state. Once that's  implemented I will approve!

**Comment by JAAlvarado-Montes** [comment]
Hi @danjampro,

I have addressed your comment as suggested. Please let me know whether you find it's ok, so I can close and merge this PR 👍 

Cheers,
Jaime.

**Review by danjampro — APPROVED** [review]
LGTM. Let's wait for either @AnthonyHorton of @wtgee to approve before the merge.

**Pull request merged**
